posthog/models/action.py

```diff
@@ -36,49 +37,42 @@ def calculate_events(self, start=None, end=None):
                 event_query, params = (
                     Event.objects.query_db_by_action(self, start=start, end=end).only("pk").query.sql_with_params()
                 )
-
         except EmptyResultSet:
-            self.last_calculated_at = calculated_at
-            self.is_calculating = False
-            self.save()
             self.events.all().delete()
-            return
-
-        query = """DELETE FROM "posthog_action_events" WHERE "action_id" = {}""".format(self.pk)
-
-        period_delete_query = """AND "event_id" in
-                       (SELECT id
-                        FROM posthog_event
-                        WHERE "created_at" >= '{}'
-                        AND "created_at" < '{}');
-                """.format(
-            start.isoformat(), end.isoformat()
-        )
-
-        insert_query = """INSERT INTO "posthog_action_events" ("action_id", "event_id")
-                        {}
-                    ON CONFLICT DO NOTHING
-                 """.format(
-            event_query.replace("SELECT ", "SELECT {}, ".format(self.pk), 1)
-        )
-
-        if not recalculate_all:
-            query += period_delete_query
         else:
-            query += ";"
-
-        query += insert_query
-
-        cursor = connection.cursor()
-        with transaction.atomic():
-            try:
-                cursor.execute(query, params)
-            except Exception as err:
-                capture_exception(err)
-
-        self.is_calculating = False
-        self.last_calculated_at = calculated_at
-        self.save()
+            delete_query = f"""DELETE FROM "posthog_action_events" WHERE "action_id" = {self.pk}"""
+
+            if not recalculate_all:
+                delete_query += f""" AND "event_id" IN (
+                    SELECT id FROM posthog_event
+                    WHERE "created_at" >= '{start.isoformat()}' AND "created_at" < '{end.isoformat()}'
+                )"""
+
+            insert_query = """INSERT INTO "posthog_action_events" ("action_id", "event_id")
+                            {}
+                        ON CONFLICT DO NOTHING
+                    """.format(
+                event_query.replace("SELECT ", f"SELECT {self.pk}, ", 1)
+            )
+
+            cursor = connection.cursor()
+            with transaction.atomic():
+                try:
+                    cursor.execute(delete_query + ";" + insert_query, params)
+                except Exception as err:
+                    capture_exception(err)
+
+            if self.post_to_slack:
+                for event in self.events.filter(
```

Comment:
This seems like it could lead to double-deliveries right?

1. time=0 calculate_events gets called, stalls before this (e.g. on the insert/delete)
2. time=30 calculate_events gets called again
3. Query from time=30 resolves, bunch of events sent to be posted for webhooks
4. Query from time=0 resolves, bunch of events get sent to be posted for webhooks.

Reply:
Also when we Insert + Delete (above) and insert again